redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+        :type res: dict
+
+        `res` should be in the following format:
+            Dict<node_name, command_result>
+        """
+        if command in self.result_callbacks:
+            return self.result_callbacks[command](command, res, **kwargs)
+        elif len(res) == 1:
+            # When we execute the command on a single node, we can
+            # remove the dictionary and return a single response
+            return list(res.values())[0]
+        else:
+            return res
+
+
+class ClusterNode(object):
+    def __init__(self, host, port, server_type=None, redis_connection=None):
+        if host == 'localhost':
+            host = socket.gethostbyname(host)
+
+        self.host = host
+        self.port = port
+        self.name = get_node_name(host, port)
+        self.server_type = server_type
+        self.redis_connection = redis_connection
+
+    def __repr__(self):
+        return '[host={0},port={1},' \
+               'name={2},server_type={3},redis_connection={4}]' \
+            .format(self.host,
+                    self.port,
+                    self.name,
+                    self.server_type,
+                    self.redis_connection)
+
+    def __eq__(self, obj):
+        return isinstance(obj, ClusterNode) and obj.name == self.name
+
+
+class LoadBalancer:
+    """
+    Round-Robin Load Balancing
+    """
+
+    def __init__(self, start_index=0):
+        self.primary_to_idx = {}
+        self.start_index = start_index
+
+    def get_server_index(self, primary, list_size):
+        server_index = self.primary_to_idx.setdefault(primary,
+                                                      self.start_index)
+        # Update the index
+        self.primary_to_idx[primary] = (server_index + 1) % list_size
+        return server_index
+
+    def reset(self):
+        self.primary_to_idx.clear()
+
+
+class NodesManager:
+    def __init__(self, startup_nodes, from_url=False,
+                 require_full_coverage=True, skip_full_coverage_check=False,
+                 lock=None, **kwargs):
+        self.nodes_cache = {}
+        self.slots_cache = {}
+        self.startup_nodes = {}
+        self.populate_startup_nodes(startup_nodes)
+        self.from_url = from_url
+        self._require_full_coverage = require_full_coverage
+        self._skip_full_coverage_check = skip_full_coverage_check
+        self._moved_exception = None
+        self.connection_kwargs = kwargs
+        self.read_load_balancer = LoadBalancer()
+        if lock is None:
+            lock = threading.Lock()
+        self._lock = lock
+        self.initialize()
+
+    def get_node(self, host=None, port=None, node_name=None):
+        if node_name is None and (host is None or port is None):
+            warnings.warn(
+                "get_node requires one of the followings: "
+                "1. node name "
+                "2. host and port"
+            )
+            return None
+        if host is not None and port is not None:
```

Comment:
Can we do something to reduce the condition checks. First we're checking the node name and the host/port combination. Then for none, then the host.